### PR TITLE
fix crash in dequantizelinear/quantizelinear for optional zero point

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.
fix the issue #4032 and #3802 in OnnxRuntime side. For the quantizeLinear, there also needs a fix in ONNX type inference. Will do that in ONNX repo.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/quantize_linear.cc [hunk 1]
```diff
@@ -26,43 +26,48 @@ ONNX_CPU_OPERATOR_TYPED_KERNEL(
...
   } else {
     N = 1;
     broadcast_dim = 1;
     block_size = static_cast<size_t>(x_shape.Size());
 
     // if no axis, enforce that scale and zero point are scalars
-    ORT_ENFORCE(IsScalarOr1ElementVector(&x_scale), "x_scale must be a scalar or 1D tensor or size 1.");
-    ORT_ENFORCE(IsScalarOr1ElementVector(&x_zero_point), "x_zero_point must be a scalar or 1D tensor or size 1.");
+    ORT_ENFORCE(IsScalarOr1ElementVector(x_scale), "x_scale must be a scalar or 1D tensor or size 1.");
+    ORT_ENFORCE( x_zero_point == nullptr || IsScalarOr1ElementVector(x_zero_point), "x_zero_point must be a scalar or 1D tensor or size 1.");
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/tensor/quantize_linear.cc:60]: nit: bad spacing after ORT_ENFORCE

File: onnxruntime/core/providers/cpu/tensor/quantize_linear.cc [hunk 2]
```diff
@@ -26,43 +26,48 @@ ONNX_CPU_OPERATOR_TYPED_KERNEL(
...
   }
 
-  const T* zero_point = x_zero_point.template Data<T>();
-  const float* scale = x_scale.template Data<float>();
-  const T* input = x.template Data<T>();
-  float* output = y.template MutableData<float>();
+  const T* zero_point = x_zero_point? x_zero_point->template Data<T>() : nullptr;
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/tensor/quantize_linear.cc:63]: doesn't clang-format add a space before the ? op?

File: onnxruntime/core/providers/cpu/tensor/quantize_linear.cc [hunk 3]
```diff
@@ -26,43 +26,48 @@ ONNX_CPU_OPERATOR_TYPED_KERNEL(
...
+  const float* scale = x_scale->template Data<float>();
+  const T* input = x->template Data<T>();
+  float* output = y->template MutableData<float>();
 
   for (size_t n = 0; n < static_cast<size_t>(N); n++) {
     for (size_t bd = 0; bd < static_cast<size_t>(broadcast_dim); bd++) {
-      auto zp = static_cast<int32_t>(zero_point[bd]);
+      auto zp = zero_point? static_cast<int32_t>(zero_point[bd]) : 0;
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/tensor/quantize_linear.cc:70]: same about the ternary op: extra space?

File: onnxruntime/core/providers/cpu/tensor/quantize_linear.cc [hunk 4]
```diff
@@ -26,43 +26,48 @@ ONNX_CPU_OPERATOR_TYPED_KERNEL(
...
-  auto& x_zero_point = *ctx->Input<Tensor>(2);
-  const auto& x_shape = x.Shape();
-  auto& y = *ctx->Output(0, x_shape);
+  auto x = ctx->Input<Tensor>(0);
+  auto x_scale = ctx->Input<Tensor>(1);
+  ORT_ENFORCE(x != nullptr && x_scale != nullptr);
+  auto x_zero_point = ctx->Input<Tensor>(2);
+
+  const auto& x_shape = x->Shape();
+  auto y = ctx->Output(0, x_shape);
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/tensor/quantize_linear.cc:35]: do "auto*" instead?

File: onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc [hunk 5]
```diff
@@ -55,6 +55,16 @@ TEST(DequantizeLinearOpTest, DequantizeLinear_Scalar) {
   test.Run();
 }
 
+// dequantize without zero point
+TEST( DequantizeLinearOpTest, DequantizeLinear_Without_Zero_Point )
```
Comment (tracysh) [inline onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc:59]: The spacing used in this test is different than the surrounding code and should be consistent.

File: onnxruntime/core/providers/cpu/tensor/quantize_linear.cc [hunk 6]
```diff
@@ -26,43 +26,48 @@ ONNX_CPU_OPERATOR_TYPED_KERNEL(
 template <typename T>
 // formula is Y = (X - ZeroPoint) * Scale
 Status DequantizeLinear<T>::Compute(OpKernelContext* ctx) const {
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/tensor/quantize_linear.cc:28]: Your change fixes DequantizeLinear for CPU, but what about CUDA EP? QuantizeLinear also has an optional zero point tensor and should be fixed up too.

Comment (yufenglee) [reply onnxruntime/core/providers/cpu/tensor/quantize_linear.cc:28]: There is a bug in onnx type inference to support QuantizeLinear with optional zero point. Will fix that later.

File: onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc (hunk 5 above)
Comment (yufenglee) [reply onnxruntime/test/providers/cpu/tensor/quantize_linear_test.cc]: forgot to run clang-format.

**Pull request merged**